Make rendering changes to make transform-style:preserve-3d and perspective only apply to direct DOM children.

mattwoodrow opened the pull request:
#### 1ad0e15b45126e8e484bbc3ecd57c2f29df74d8c
<pre>
Make rendering changes to make transform-style:preserve-3d and perspective only apply to direct DOM children.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248282">https://bugs.webkit.org/show_bug.cgi?id=248282</a>
&lt;rdar://problem/102632581&gt;

Reviewed by Simon Fraser.

The current editors draft for css-transforms-2 requires that transform-style:preserve-3d and perspective only apply to
direct DOM children elements. The current behaviour is that they are applied to the nearest descendants that result in
creation of a layer (which isn&apos;t very specific to the WebKit implementation).

This inserts a spacer GraphicsLayer on a child RenderLayer that would inherit the transform from its parent layer, but shouldn&apos;t due
to it not belong to the direct DOM parent element.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/preserve3d-and-flattening-001.html:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::useCSS3DTransformInteroperability const):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::hasPerspective const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::destroyGraphicsLayers):
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateInternalHierarchy):
(WebCore::ancestorLayerIsDOMParent):
(WebCore::ancestorLayerWillCombineTransform):
(WebCore::RenderLayerBacking::updateTransformSpacerLayer):
(WebCore::RenderLayerBacking::childForSuperlayers const):
(WebCore::RenderLayerBacking::useCSS3DTransformInteroperability const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::recompositeChangeRequiresChildrenGeometryUpdate):
(WebCore::RenderLayerCompositor::layerStyleChanged):

Canonical link: <a href="https://commits.webkit.org/257190@main">https://commits.webkit.org/257190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91b3d2d91a3ade17f5f4b4d961bfc0b5b2415365

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98080 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107542 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167808 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102021 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7783 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36058 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90696 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104152 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103732 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5854 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84678 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32994 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89439 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1269 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1233 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22375 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4952 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6094 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2534 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41780 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerBacking.h:
```diff
@@ -419,6 +421,7 @@ class RenderLayerBacking final : public GraphicsLayerClient {
     RefPtr<GraphicsLayer> m_childContainmentLayer; // Only used if we have clipping on a stacking context with compositing children, or if the layer has a tile cache.
     RefPtr<GraphicsLayer> m_viewportAnchorLayer; // Only used on fixed/sticky elements.
     RefPtr<GraphicsLayer> m_maskLayer; // Only used if we have a mask and/or clip-path.
+    RefPtr<GraphicsLayer> m_transformSpacerLayer;
```
"Spacer" doesn't really communicate this layer's role (it's not adding space). `m_transformFlatteningLayer` or just `m_flatteningLayer`?

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2241,6 +2248,45 @@ void RenderLayerBacking::positionOverflowControlsLayers()
+
+bool RenderLayerBacking::updateTransformSpacerLayer(const RenderLayer* compositingAncestor)
+{
+    bool needsSpacerLayer = false;
```
spacer -> flattening

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2241,6 +2248,45 @@ void RenderLayerBacking::positionOverflowControlsLayers()
+    bool layerChanged = false;
+    if (needsSpacerLayer) {
+        if (!m_transformSpacerLayer) {
+            String layerName = makeString(m_owningLayer.name(), " (3d spacer)");
```
no need for the local variable.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1792,6 +1798,8 @@ void RenderLayerCompositor::layerStyleChanged(StyleDifference diff, RenderLayer&
             layer.setNeedsCompositingGeometryUpdate();
         }
+        if (m_renderView.settings().css3DTransformInteroperabilityEnabled() && oldStyle && recompositeChangeRequiresChildrenGeometryUpdate(*oldStyle, newStyle))
+            layer.setChildrenNeedCompositingGeometryUpdate();
```
If the flattening layer comes and goes, this is a "configuration update", not a "geometry update".

mattwoodrow replied:
It appears that `setChildrenNeedCompositingConfigurationUpdate` doesn't exist, and `setChildrenNeedCompositingGeometryUpdate` implies a configuration update as well as a geometry update.

mattwoodrow committed:
Make rendering changes to make transform-style:preserve-3d and perspective only apply to direct DOM children.
https://bugs.webkit.org/show_bug.cgi?id=248282
<rdar://problem/102632581>

Reviewed by Simon Fraser.

The current editors draft for css-transforms-2 requires that transform-style:preserve-3d and perspective only apply to
direct DOM children elements. The current behaviour is that they are applied to the nearest descendants that result in
creation of a layer (which isn't very specific to the WebKit implementation).

This inserts a spacer GraphicsLayer on a child RenderLayer that would inherit the transform from its parent layer, but shouldn't due
to it not belong to the direct DOM parent element.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/preserve3d-and-flattening-001.html:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::useCSS3DTransformInteroperability const):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::hasPerspective const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::destroyGraphicsLayers):
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateInternalHierarchy):
(WebCore::ancestorLayerIsDOMParent):
(WebCore::ancestorLayerWillCombineTransform):
(WebCore::RenderLayerBacking::updateTransformSpacerLayer):
(WebCore::RenderLayerBacking::childForSuperlayers const):
(WebCore::RenderLayerBacking::useCSS3DTransformInteroperability const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::recompositeChangeRequiresChildrenGeometryUpdate):
(WebCore::RenderLayerCompositor::layerStyleChanged):

Canonical link: https://commits.webkit.org/257190@main
